Bump click component for bump mining

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

https://github.com/tgstation/tgstation/pull/64221 without touching firelocks because @LemonInTheDark said they liked the component and I put a lot of work into it. 

Adds the component to bump mining turfs and removes the code under Bumped() that handled it.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Cuts down on copypasted code if somebody wants to add bump clicking to something else in future.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->
:cl:
balance: Bump mining requires activehandedness
code: Bump click element for when bumping something should click it, like mineral turfs
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Tool_bump component

LemonInTheDark commented on code/datums/elements/tool_bump.dm:
```diff
@@ -0,0 +1,87 @@
+
+	if(!isatom(target) || isarea(target))
+		return ELEMENT_INCOMPATIBLE
+	LAZYADD(src.tool_behaviours, tool_behaviours)
+	src.require_active_hand = require_active_hand
+	LAZYADD(src.tool_items, tool_items)
```
Why make these lists lazy? couldn't you just set them to be equal to the args?

cacogen replied:
I assumed that having a list on every mining turf would be bad for performance and that making it a lazy list would be better. Is it different for elements? I'm not sure what you mean by set them to be equal to the args. Just `src.tool_items += tool_items`?

LemonInTheDark replied:
A bespoke element like the one you've created is different from a component. 
For each set of arguments, a unique element is created.
So if you add an element twice with the same args, no duplication happens.

Regardless of that. tool_behaviors and tool_items default to null, so why not just set src.tool_behaviors and src.tool_items directly? All a non filled lazy list is is a null value anyhow

cacogen replied:
Okay, done

LemonInTheDark commented on code/datums/elements/tool_bump.dm:
```diff
@@ -0,0 +1,87 @@
+
+	var/obj/item/held_tool
+	if(iscyborg(bumper))
+		var/mob/living/silicon/robot/cyborg = bumper
+
+		held_tool = cyborg.module_active
+		if(!held_tool)
+			return
+
+		if(LAZYFIND(tool_behaviours, held_tool.tool_behaviour))
+			INVOKE_ASYNC(held_tool, /obj/item.proc/melee_attack_chain, bumper, source)
+			return
+		for(var/tool_path in tool_items)
+			if(istype(held_tool, tool_path))
+				INVOKE_ASYNC(held_tool, /obj/item.proc/melee_attack_chain, bumper, source)
+		return
+
+	if(require_active_hand)
+		held_tool = bumper.get_active_held_item()
+		if(!held_tool)
+			return
+		if(LAZYFIND(tool_behaviours, held_tool.tool_behaviour))
+			INVOKE_ASYNC(held_tool, /obj/item.proc/melee_attack_chain, bumper, source)
+			return
+		for(var/tool_path in tool_items)
+			if(istype(held_tool, tool_path))
+				INVOKE_ASYNC(held_tool, /obj/item.proc/melee_attack_chain, bumper, source)
+				return
+	else
+		for(var/tool_behaviour in tool_behaviours)
+			held_tool = bumper.is_holding_tool_quality(tool_behaviour)
+			if(!held_tool)
+				continue
+			INVOKE_ASYNC(held_tool, /obj/item.proc/melee_attack_chain, bumper, source)
+			return
+		for(var/tool_path in tool_items)
+			for(held_tool in bumper.held_items)
+				if(istype(held_tool, tool_path))
+					INVOKE_ASYNC(held_tool, /obj/item.proc/melee_attack_chain, bumper, source)
+					return
```
It feels like you were planning on designing this to reduce copypaste, and then stopped. Was that just because of the non active hand bit?

What we want is basically "find a valid tool, then call the melee attack chain with it if we find one" right?

Could you make a proc that returns the tool to use to help make this cleaner? or would that not work

cacogen replied:
I guess I could try to cut this down into less lines but I guess I don't understand the point of it.

LemonInTheDark commented on code/game/turfs/closed/minerals.dm:
```diff
@@ -36,7 +36,7 @@
 	transform = M
 	icon = smooth_icon
-
+	AddElement(/datum/element/tool_bump, tool_behaviours = list(TOOL_MINING), require_active_hand = FALSE)
```
Bespoke elements work based off checking if any existing elements of that type share the same set of arguments.
Unfortunately, they don't care about the contents of a list, they only care about the EXACT same list
So for elements like this we pass in locally static lists, to ensure the memory benefits are actually felt 
```suggestion
	var/static/list/behaviors = list(TOOL_MINING)
	AddElement(/datum/element/tool_bump, tool_behaviours = behaviors, require_active_hand = FALSE)
```

cacogen committed:
Cuts down on code, incorporates Lemon's feedback

Fikou commented:
Does this work with concussive gauntlets?

cacogen committed:
Bump click now

cacogen commented:
>Does this work with concussive gauntlets?

It has an unusual interaction where the concussive gauntlets mine the mineral turf and then if you were holding a mining tool when you bumped it it's then used to dig the ground where the mineral turf was. Not sure how to prevent that and don't think it really matters. If you immediately move into the rockface you mined it doesn't complete anyway.

cacogen commented:
@LemonInTheDark I don't know if you want this now, I worked my way all the way up the attack chain from it initially calling `attackby()` to it then calling `melee_attack_chain()` to it now finally calling `ClickOn()`. 

The first change was to make it work with resonators (i.e. call `pre_attack()`) and the second was to make it work with hand mining turfs (i.e. call `attack_hand()`). 

I don't think players are going to like it being on bump mining because making it click means it requires activehandedness. I don't have any use-cases in mind for when it would actually matter though.

LemonInTheDark commented on code/datums/elements/bump_click.dm:
```diff
@@ -0,0 +1,75 @@
+		if(allow_unarmed)
+			INVOKE_ASYNC(bumper, /mob.proc/ClickOn, source) //Click with empty active hand
+		return
+	if(!ISADVANCEDTOOLUSER(bumper)) //probably can't perform this interaction anyway
```
did cyborg code get dusted? or what does you have now support them

LemonInTheDark commented on code/datums/elements/bump_click.dm:
```diff
@@ -0,0 +1,75 @@
+			continue
+		return held_tool
+
+	for(held_tool in bumper.held_items)
```
sorry about this, I mislead you a bit in code suggestions I think.
You're gonna want to avoid this pattern of passing in existing vars into for loops. It works, but it makes code harder to track, and can be a bit confusing

LemonInTheDark commented on code/datums/elements/bump_click.dm:
```diff
@@ -0,0 +1,75 @@
+	if(!isatom(target) || isarea(target))
+		return ELEMENT_INCOMPATIBLE
+	src.tool_behaviours = tool_behaviours
+	src.tool_items = typecacheof(tool_items)
```
Typecaches are expensive to generate. s important we only create them once, if we need them, rather then doing it each time the element is added
```suggestion
	if(length(tool_items) && !length(src.tool_items))
		src.tool_items = typecacheof(tool_items)
```

LemonInTheDark commented:
I'll look into it a bit deeper. If you think it's reasonably I trust your judgement

Mothblocks commented on code/datums/elements/bump_click.dm:
```diff
@@ -0,0 +1,75 @@
+ * is a specific item or its subtypes or if the active hand slot is empty and empty-handedness is allowed.
+ */
+/datum/element/bump_click
+	element_flags = ELEMENT_BESPOKE
```
```suggestion
	element_flags = ELEMENT_BESPOKE | ELEMENT_DETACH
```

Mothblocks commented on tgstation.dme:
```diff
@@ -953,6 +953,7 @@
 #include "code\datums\elements\strippable.dm"
 #include "code\datums\elements\swabbable.dm"
 #include "code\datums\elements\tenacious.dm"
+#include "code\datums\elements\bump_click.dm"
```
Not in ABC order, use "ticked" in the bottom right of VSC

cacogen committed:
Merge remote-tracking branch 'origin/master' into toolbumpcomponent

cacogen committed:
Completely removes support for off-hand items

Bumping switching hands for you seems like a bad idea when fighting mining mobs. 

You can imagine brushing a mining turf while fleeing a mining mob with a mining item in your off-hand and the weapon you're using to fight it in your active hand and the switch being something you don't immediately catch that stops a couple of your attacks from going through and possibly gets you killed.

cacogen committed:
Let the things being clicked decide whether you need advanced tool use

cacogen committed:
Needs to check for living

cacogen committed:
Updates comments

tohg7 commented:
Will this work with Intelli-monkeys? (helmeted sentient monkeys)

cacogen commented:
>Will this work with Intelli-monkeys? (helmeted sentient monkeys)

The component itself works for all living mobs (should probably check for a mind, in hindsight). Don't think monkeys can mine with tools because as far as I can tell they aren't advanced tool users.
